spec/unit/pops/types/p_type_set_type_spec.rb

```diff
@@ -0,0 +1,424 @@
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                /entry 'references' entry 'Ref' expected a Struct value, got Integer/)
+            end
+
+            it 'contains several initialization that refers to the same TypeSet' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                references => {
+                  A => { name => 'Vehicle::Cars', version_range => '1.x' },
+                  V => { name => 'Vehicle::Cars', version_range => '1.x' },
+                }
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(ArgumentError,
+                /references TypeSet 'http:\/\/puppet\.com\/2016\.1\/runtime\/Vehicle::Cars'/)
+            end
+
+            it 'contains an initialization maps with an alias that collides with a type name' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                types => {
+                  Car => Object[{}]
+                },
+                references => {
+                  Car => { name => 'Vehicle::Car', version_range => '1.x' }
+                }
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(ArgumentError,
+                /references a TypeSet using alias 'Car'. The alias collides with the name of a declared type/)
+            end
+
+            context 'contains an initialization map that' do
+              it 'has no version range' do
+                ts = <<-OBJECT
+                  pcore_version => '1.0.0',
+                  version => '1.0.0',
+                  references => { Ref => { name => 'X' } }
+                OBJECT
+                expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                  /entry 'references' entry 'Ref' expected a value for key 'version_range'/)
+              end
+
+              it 'has no name' do
+                ts = <<-OBJECT
+                  pcore_version => '1.0.0',
+                  version => '1.0.0',
+                  references => { Ref => { version_range => '1.x' } }
+                OBJECT
+                expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                  /entry 'references' entry 'Ref' expected a value for key 'name'/)
+              end
+
+              it 'has a name that is not a QRef' do
+                ts = <<-OBJECT
+                  pcore_version => '1.0.0',
+                  version => '1.0.0',
+                  references => { Ref => { name => 'cars', version_range => '1.x' } }
+                OBJECT
+                expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                  /entry 'references' entry 'Ref' entry 'name' expected a match for Pattern\[\/\\A\(\(::\)\{0,1\}\[A-Z\]\[\\w\]\*\)\+\\z\/\], got 'cars'/)
```

Comment:
Do we allow leading `::` in names ? That would allow breaking out of the type set namespace and reference anything anywhere without having a reference to it.
Am I reading this wrong?